src/tasks.ts

```diff
@@ -0,0 +1,113 @@
+import { spawnSync } from "child_process";
+import * as path from "path";
+import { Range, TextEditor } from "vscode";
+import { outputChannel, ws } from "./extension";
+import { TaskOutput } from "./taskRunner";
+import { NameToThemeNumber } from "./themePicker";
+import { util } from "./utility";
+
+/**
+ * D2Tasks - static functions to be used in tasks.  Functions need
+ * to be synchronous, and return their results in a callback.
+ */
+class D2Tasks {
+  public compile(
+    text: string,
+    filePath?: string,
+    log?: TaskOutput,
+    terminal?: TaskOutput
+  ): string {
+    const layout: string = ws.get("previewLayout", "dagre");
+    const theme: string = ws.get("previewTheme", "default");
+    const sketch: boolean = ws.get("previewSketch", false);
+    const themeNumber: number = NameToThemeNumber(theme);
+    const d2Path: string = ws.get("execPath", "d2");
+
+    terminal?.("Starting Compile...");
+    terminal?.(`Layout: ${layout}  Theme: ${theme}  Sketch: ${sketch}`);
+    terminal?.(`Current Working Directory: ${filePath}`);
+    terminal?.("");
+
+    const args: string[] = [
+      `--layout=${layout}`,
+      `--theme=${themeNumber}`,
+      `--sketch=${sketch}`,
+      "-",
+    ];
+    const proc = spawnSync(d2Path, args, {
+      cwd: filePath,
+      input: text,
+      encoding: "utf-8",
+    });
+
+    /** proc.status: 0 - success
+     *  proc.status: 1 - errors
+     *  proc.pid: 0 - EXE Not Found
+     */
+    if (proc.pid === 0) {
+      util.showErrorToolsNotFound(proc.error?.message ?? "");
+    } else {
+      const error: string = proc.stderr?.toString() ?? "";
+
+      for (const msg of error.split("\n")) {
+        if (msg.length === 0) {
+          // Sometimes there is a blank line that we
+          // don't have to worry about
+          continue;
+        }
+
+        terminal?.(msg, true);
+      }
+    }
+
+    terminal?.("");
+
+    let data = "";
+    // Success! Read the output
+    if (proc.status === 0) {
```

Comment:
```suggestion
    if (!proc.status) {
```

I'm getting status `null` for success and without this suggestion the preview never loads for me

Reply:
Which operating system?  Also, what is the value of proc.signal when this happens?  A null value in status says that the process terminated with a signal, which would be odd if it was successful as you said.

Reply:
interesting this is proc and it has an error, but also has the correct output and no stderr:

```
"error":{
  "errno":-55,
  "code":"ENOBUFS",
  "syscall":"spawnSync d2",
  "path":"d2",
  "spawnargs":["--layout=tala","--theme=3","--sketch=false","-"]
},
"status":null,
"signal":"SIGPIPE",
"output":[null,"<?xml version=\"1.0\" encoding=\"utf-8\"?>\n<svg\nid=\"d2-svg\"\ ... rest of svg ... ",
"stderr":""
}
```

this is on macos btw

Reply:
the above is on https://github.com/terrastruct/d2/blob/master/docs/examples/twitter/in.d2#L0-L1
on a smaller diagram with fewer images (https://github.com/terrastruct/d2/blob/master/docs/examples/chess/dia.d2#L0-L1)

proc has the following: 

```
{
"status":0,
"signal":null,
"output":[null,"<?xml version=\"1.0\" encoding=\"utf-8\"?>\n<svg\nid=\"d2-svg\" ... rest of svg ... ",
"stderr":"success: successfully compiled - to -\n"
}
```

Reply:
ok I fixed it by increasing the maxbuffer size `      maxBuffer: 1024 * 1024 * 1024,` in `spawnSync`.

Reply:
I was able to reproduce this on Linux, so I reworked some of the error code and set the buffer size to 24meg.  Your example was 88k above the 1M limit.  Since the font data comes last, just letting it succeed probably would render correctly even with the font data truncated.

I picked 24meg as the default since font data makes up about 98% of the examples returned data, so you'd have to have the worlds most complicated chart to even make it to two meg.  Now, on the other hand, if multiple fonts are/can be specified, then you'd still have to have *a lot* of fonts to push this boundary.  If that becomes the case, then this could be an option that can be set.

While 1gig is fine on a 64bit system, if vscode is run on a 32bit system, a free contiguous 1gig might be hard to come by.  At least node only commits the memory as needed instead of all at once.

Reply:
makes sense 1 gig is indeed a bit excessive